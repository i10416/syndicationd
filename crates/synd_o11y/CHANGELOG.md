# Changelog

All notable changes to this project will be documented in this file.

## [v0.1.7] 2024-06-18

### Miscellaneous Tasks

- Update opentelemetry from 0.22.0 to 0.23.0 by [@ymgyt](https://github.com/ymgyt) ([1f291fc3](https://github.com/ymgyt/syndicationd/commit/1f291fc31ec07d5f84565518d848b0822cccb879))

### Refactor

- Remove empty line by [@ymgyt](https://github.com/ymgyt) ([ebafc736](https://github.com/ymgyt/syndicationd/commit/ebafc7364333e212bbd56e5e536c7717a96c749b))

### Testing

- Add tracing opentelemetry layer tests (#50) by [@Yuta Yamaguchi](https://github.com/Yuta Yamaguchi) ([9739b518](https://github.com/ymgyt/syndicationd/commit/9739b5189bdeaedd458771a92a04e416f1af08fd))
- Add metrics view test case by [@ymgyt](https://github.com/ymgyt) ([9384adf3](https://github.com/ymgyt/syndicationd/commit/9384adf336db680e50a380ab2f42b4113108f691))
- Add http extractor test case by [@ymgyt](https://github.com/ymgyt) ([87cbd5f1](https://github.com/ymgyt/syndicationd/commit/87cbd5f1d817f0e48f570b895d6c533b99172e30))

## [v0.1.6] - 2024-05-06

### Documentation

- Add comments to public api by [@ymgyt](https://github.com/ymgyt) ([41588384](https://github.com/ymgyt/syndicationd/commit/41588384a0344a3befc473b9ca45abe6be2054ac))

## [v0.1.5] - 2024-04-17

### Features

- Add opentelemetry-stdout feature by [@ymgyt](https://github.com/ymgyt) ([765c036d](https://github.com/ymgyt/syndicationd/commit/765c036dc143f976e108935943bd9f89f03deea7))

### Miscellaneous Tasks

- Change view event level to debug by [@ymgyt](https://github.com/ymgyt) ([1e8e0c07](https://github.com/ymgyt/syndicationd/commit/1e8e0c07cffd48dc42e202185f42a0afdfd2fd03))

### Refactor

- Compose otel layer by [@ymgyt](https://github.com/ymgyt) ([3562f3f0](https://github.com/ymgyt/syndicationd/commit/3562f3f0eb6224f89181ab6af87ec0b1c2e2403c))
- Use semantic conventions const by [@ymgyt](https://github.com/ymgyt) ([2f1bac7d](https://github.com/ymgyt/syndicationd/commit/2f1bac7d407d32078ce48ef8a495fa2deecbb9d2))

## [v0.1.4] - 2024-03-08

### Miscellaneous Tasks

- Disable cargo-dist by [@ymgyt](https://github.com/ymgyt) ([d7cf038f](https://github.com/ymgyt/syndicationd/commit/d7cf038f329f43645d49667b923125879afb8e1c))
- Remove workspace otel deps by [@ymgyt](https://github.com/ymgyt) ([20c071ea](https://github.com/ymgyt/syndicationd/commit/20c071ea08d1c4afc1cd0a724037bbdfa10eb1cf))
- Update opentelemetry from 0.21 to 0.22 by [@ymgyt](https://github.com/ymgyt) ([b874c1d5](https://github.com/ymgyt/syndicationd/commit/b874c1d5ba804339a495a29a28cfd6443b2e2339))

## [v0.1.3] - 2024-02-25

### Features

- Add health check response type by [@ymgyt](https://github.com/ymgyt) ([f2420630](https://github.com/ymgyt/syndicationd/commit/f242063027b2ba5cac06a871a4c24d2413366cf4))
- Impl fmt::Display for health check status by [@ymgyt](https://github.com/ymgyt) ([b52bd8d5](https://github.com/ymgyt/syndicationd/commit/b52bd8d56d39bc4263f0c4851fb078803bd65881))

### Miscellaneous Tasks

- Trim prefix from changelog by [@ymgyt](https://github.com/ymgyt) ([95d44877](https://github.com/ymgyt/syndicationd/commit/95d448773ec7ab009fbece0928854364679b6f2c))
- Add homepage to package metadata by [@ymgyt](https://github.com/ymgyt) ([4bfdb49e](https://github.com/ymgyt/syndicationd/commit/4bfdb49e317e18ff6345ce1b8e8071f0497a1a5f))

## [v0.1.2] - 2024-02-19

### Features

- Instrument request counter metrics by [@ymgyt](https://github.com/ymgyt) ([ac64b3aa](https://github.com/ymgyt/syndicationd/commit/ac64b3aa6880482597e672649de800eb30b3ad56))
- Add metric macro by [@ymgyt](https://github.com/ymgyt) ([5b723f4b](https://github.com/ymgyt/syndicationd/commit/5b723f4b0c68b422f4778b502d2136ef4662bebd))
- Configure endpoint by [@ymgyt](https://github.com/ymgyt) ([14e09265](https://github.com/ymgyt/syndicationd/commit/14e0926596c59a5e32c283d0f8ac7f805e9e97d9))

### Documentation

- Fix typo by [@ymgyt](https://github.com/ymgyt) ([c311417b](https://github.com/ymgyt/syndicationd/commit/c311417bb69d22c7826d4ec931ec8dfe59042ca1))

## [v0.1.1] - 2024-02-12

### Features

- Add o11y crate by [@ymgyt](https://github.com/ymgyt) ([0a50517e](https://github.com/ymgyt/syndicationd/commit/0a50517e0b861973fac95ad5dba6f2c4d5b7270d))
- Add opentelemetry-tracing log bridge layer by [@ymgyt](https://github.com/ymgyt) ([92f22b56](https://github.com/ymgyt/syndicationd/commit/92f22b564357a0d43f8631212cf976338eb05a04))
- Add baggage propagation by [@ymgyt](https://github.com/ymgyt) ([d02e514c](https://github.com/ymgyt/syndicationd/commit/d02e514c8f6e32aa748c10dadb204153cba21ecc))
- Add opentelemetry layers by [@ymgyt](https://github.com/ymgyt) ([4d3f5bf3](https://github.com/ymgyt/syndicationd/commit/4d3f5bf3f45f31cfd014dbdf37a41a31ea0472ca))

### Miscellaneous Tasks

- Configure release flow by [@ymgyt](https://github.com/ymgyt) ([855d1063](https://github.com/ymgyt/syndicationd/commit/855d1063f5b476433fe0a7ab352b72d63a749e2e))
- Use hyphen as package name instead of underscore by [@ymgyt](https://github.com/ymgyt) ([0a8ed059](https://github.com/ymgyt/syndicationd/commit/0a8ed05997790f9f05c932c92fa2b2b2d74065a9))

### Testing

- Fix wrong assertion by [@ymgyt](https://github.com/ymgyt) ([392df0de](https://github.com/ymgyt/syndicationd/commit/392df0de6c4b7e8c34ae2d7c2f8ec764c23145f0))

<!-- generated by git-cliff -->
