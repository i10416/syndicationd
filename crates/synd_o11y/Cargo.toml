[package]
authors.workspace    = true
categories.workspace = true
edition.workspace    = true
keywords.workspace   = true
license.workspace    = true
readme.workspace     = true
repository.workspace = true

description = "syndicationd lib for tracing and opentelemetry ecosystem"
name        = "synd_o11y"
version     = "0.1.0"

[dependencies]
# disable logs_level_enabled which affect global filitering
opentelemetry                      = { workspace = true }
opentelemetry-appender-tracing     = { version = "0.2.0", default-features = false }
opentelemetry-otlp                 = { version = "0.14.0", default-features = false, features = ["trace", "metrics", "logs", "grpc-tonic"] }
opentelemetry-semantic-conventions = { workspace = true }
opentelemetry_sdk                  = { workspace = true, features = ["logs", "rt-tokio"] }
tracing                            = { workspace = true }
tracing-subscriber                 = { workspace = true }

[lints]
workspace = true
