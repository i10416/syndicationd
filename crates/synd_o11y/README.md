# synd_o11y

syndicationd observability library
