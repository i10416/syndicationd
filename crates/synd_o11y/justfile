# List recipe
default:
    just --list

# Generate CHANGELOG for given version release 
changelog version:
    git cliff \
        --config ../../cliff.toml \
        --repository ../../ \
        --include-path crates/synd_o11y/** \
        --tag-pattern synd-o11y-v.* \
        --tag synd-o11y-v{{version}} \
        --output CHANGELOG.md

# Release crate
[no-cd]
release *flags:
    cargo release --package synd-o11y -v {{flags}}
