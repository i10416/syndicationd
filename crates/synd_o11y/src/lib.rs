use ::opentelemetry::KeyValue;

pub mod opentelemetry;
pub mod tracing_subscriber;

/// Request id key for opentelemetry baggage
pub const REQUEST_ID_KEY: &str = "request.id";

/// Generate random request id
pub fn request_id() -> String {
    // https://stackoverflow.com/questions/54275459/how-do-i-create-a-random-string-by-sampling-from-alphanumeric-characters
    use rand::distributions::{Alphanumeric, DistString};
    Alphanumeric.sample_string(&mut rand::thread_rng(), 10)
}

/// Generate random request id key value
pub fn request_id_key_value() -> KeyValue {
    KeyValue::new(REQUEST_ID_KEY, request_id())
}
