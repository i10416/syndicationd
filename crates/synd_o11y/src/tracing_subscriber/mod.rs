use std::borrow::Cow;

use tracing::Subscriber;
use tracing_subscriber::{registry::LookupSpan, Layer};

pub mod audit;
pub mod otel_log;
pub mod otel_metrics;
pub mod otel_trace;

/// Return a composed layer that exports traces, metrics, and logs of opentelemetry.
/// The exported telemetry will be accompanied by a `Resource`.
pub fn opentelemetry_layer<S>(
    endpoint: impl Into<String>,
    service_name: impl Into<Cow<'static, str>>,
    service_version: impl Into<Cow<'static, str>>,
    trace_sampler_ratio: f64,
) -> impl Layer<S>
where
    S: Subscriber + for<'span> LookupSpan<'span>,
{
    let endpoint = endpoint.into();
    let resource = crate::opentelemetry::resource(service_name, service_version);

    otel_trace::layer(endpoint.clone(), resource.clone(), trace_sampler_ratio)
        .and_then(otel_metrics::layer(endpoint.clone(), resource.clone()))
        .and_then(otel_log::layer(endpoint, resource))
}
