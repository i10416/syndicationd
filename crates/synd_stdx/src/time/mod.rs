#[cfg(feature = "humantime")]
pub mod humantime;
