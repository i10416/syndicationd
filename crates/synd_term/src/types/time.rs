pub type Time = synd_feed::types::Time;

pub trait TimeExt {
    fn local_ymd(&self) -> String;
}

impl TimeExt for Time {
    fn local_ymd(&self) -> String {
        self.naive_local().format("%Y/%m/%d").to_string()
    }
}
