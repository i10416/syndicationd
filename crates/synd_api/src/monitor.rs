use std::time::Duration;

use synd_o11y::metric;
use tokio_metrics::{RuntimeMetrics, RuntimeMonitor, TaskMetrics, TaskMonitor};
use tokio_util::sync::CancellationToken;

struct Metrics {
    runtime_total_polls_count: u64,
    runtime_busy_duration_secs: f64,
    gql_mean_poll_duration_secs: f64,
    gql_mean_slow_poll_duration_secs: f64,
    gql_mean_first_poll_delay_secs: f64,
    gql_mean_scheduled_duration_secs: f64,
    gql_mean_idle_duration_secs: f64,
}

pub struct Monitors {
    gql: TaskMonitor,
}

impl Monitors {
    pub fn new() -> Self {
        Self {
            gql: TaskMonitor::new(),
        }
    }

    pub(crate) fn graphql_task_monitor(&self) -> TaskMonitor {
        self.gql.clone()
    }

    pub async fn emit_metrics(self, interval: Duration, ct: CancellationToken) {
        let handle = tokio::runtime::Handle::current();
        let runtime_monitor = RuntimeMonitor::new(&handle);
        let intervals = runtime_monitor.intervals().zip(self.gql.intervals());

        for (runtime_metrics, gql_metrics) in intervals {
            let Metrics {
                runtime_total_polls_count,
                runtime_busy_duration_secs,
                gql_mean_poll_duration_secs,
                gql_mean_slow_poll_duration_secs,
                gql_mean_first_poll_delay_secs,
                gql_mean_scheduled_duration_secs,
                gql_mean_idle_duration_secs,
            } = Self::collect_metrics(&runtime_metrics, &gql_metrics);

            // Runtime metrics
            metric!(monotonic_counter.runtime.poll = runtime_total_polls_count);
            metric!(monotonic_counter.runtime.busy_duration = runtime_busy_duration_secs);

            // Tasks poll metrics
            metric!(
                monotonic_counter.task.graphql.mean_poll_duration = gql_mean_poll_duration_secs
            );
            metric!(
                monotonic_counter.task.graphql.mean_slow_poll_duration =
                    gql_mean_slow_poll_duration_secs
            );

            // Tasks schedule metrics
            metric!(
                monotonic_counter.task.graphql.mean_first_poll_delay =
                    gql_mean_first_poll_delay_secs,
            );
            metric!(
                monotonic_counter.task.graphql.mean_scheduled_duration =
                    gql_mean_scheduled_duration_secs,
            );

            // Tasks idle metrics
            metric!(
                monotonic_counter.task.graphql.mean_idle_duration = gql_mean_idle_duration_secs,
            );

            tokio::select! {
                biased;
                // Make sure to respect cancellation
                () = ct.cancelled() => break,
                () = tokio::time::sleep(interval) => continue
            }
        }
    }

    fn collect_metrics(runtime_metrics: &RuntimeMetrics, gql_metrics: &TaskMetrics) -> Metrics {
        Metrics {
            runtime_total_polls_count: runtime_metrics.total_polls_count,
            runtime_busy_duration_secs: runtime_metrics.total_busy_duration.as_secs_f64(),
            gql_mean_poll_duration_secs: gql_metrics.mean_poll_duration().as_secs_f64(),
            gql_mean_slow_poll_duration_secs: gql_metrics.mean_slow_poll_duration().as_secs_f64(),
            gql_mean_first_poll_delay_secs: gql_metrics.mean_first_poll_delay().as_secs_f64(),
            gql_mean_scheduled_duration_secs: gql_metrics.mean_scheduled_duration().as_secs_f64(),
            gql_mean_idle_duration_secs: gql_metrics.mean_idle_duration().as_secs_f64(),
        }
    }
}

#[cfg(test)]
mod tests {
    use std::sync::{Arc, Mutex};

    use tracing::{instrument::WithSubscriber, Event, Subscriber};
    use tracing_subscriber::{
        layer::{Context, SubscriberExt as _},
        registry::LookupSpan,
        Layer, Registry,
    };

    use super::*;

    struct TestLayer<F> {
        on_event: F,
    }

    impl<S, F> Layer<S> for TestLayer<F>
    where
        S: Subscriber + for<'span> LookupSpan<'span>,
        F: Fn(&Event<'_>) + 'static,
    {
        fn on_event(&self, event: &Event<'_>, _ctx: Context<'_, S>) {
            (self.on_event)(event);
        }
    }

    #[tokio::test]
    async fn emit_metrics() {
        let events = Arc::new(Mutex::new(Vec::new()));
        let events_cloned = events.clone();
        let on_event = move |event: &Event<'_>| {
            let field = event.fields().next().unwrap().name();
            events_cloned.lock().unwrap().push(field);
        };
        let layer = TestLayer { on_event };
        let registry = Registry::default().with(layer);
        let m = Monitors::new();
        let ct = CancellationToken::new();
        ct.cancel();

        m.emit_metrics(Duration::from_millis(0), ct)
            .with_subscriber(registry)
            .await;

        let events = events.lock().unwrap().clone();
        insta::with_settings!({
            description => "metrics which monitor emits",
            omit_expression => true ,
        }, {
            insta::assert_yaml_snapshot!(events);
        });
    }
}
