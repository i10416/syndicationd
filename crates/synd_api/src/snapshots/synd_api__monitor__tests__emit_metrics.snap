---
source: crates/synd_api/src/monitor.rs
description: metrics which monitor emits
---
- monotonic_counter.runtime.poll
- monotonic_counter.runtime.busy_duration
- monotonic_counter.task.graphql.mean_poll_duration
- monotonic_counter.task.graphql.mean_slow_poll_duration
- monotonic_counter.task.graphql.mean_first_poll_delay
- monotonic_counter.task.graphql.mean_scheduled_duration
- monotonic_counter.task.graphql.mean_idle_duration
